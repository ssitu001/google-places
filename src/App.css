.App {
  text-align: center;
  /* min-height: 100vh;
  background: blue; */
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.map {
  width: 100%;
  height: 500px;
}

.row-margin {
  margin: 25px;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
