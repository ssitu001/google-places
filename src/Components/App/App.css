.app {
  text-align: center;
}

.app-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
  margin-bottom: 10px;
}

.app-logo-static {
  height: 80px;
  margin-bottom: 10px;
}

.app-header {
  background-color: #53626b;
  height: 200px;
  padding: 20px;
  color: white;
}

.app-title {
  font-size: 1.5em;
}

.app-intro {
  font-size: large;
}

.init-loading {
  margin-top: 50px;
}

.map {
  width: 100%;
  height: calc(100vh - 200px);
}

.row-margin {
  margin: 25px;
}

.no-padding {
  padding: 0;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

