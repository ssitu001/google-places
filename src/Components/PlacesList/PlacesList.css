.places-content {
  height: calc(100vh - 200px);
  flex: 1;
  overflow-y: scroll;
  cursor: pointer;
}