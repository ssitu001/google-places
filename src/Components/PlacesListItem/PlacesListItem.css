.list-group-item-alternate {
  background: #ececec;
}