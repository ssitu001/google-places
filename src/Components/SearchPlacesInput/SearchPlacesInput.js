import React from 'react';

const SearchPlacesInput = ({handleSubmit, handleChange}) => {

  return (
    <div className="row">
      <div className="col-md-12">
        <form className="form-inline" onSubmit={handleSubmit}>
          <input type="text"
            className="form-control input-md"
            onChange={handleChange}
            id="place"
            placeholder="Search"
            required/>
        </form>
      </div>
    </div>
  )
}

export default SearchPlacesInput;