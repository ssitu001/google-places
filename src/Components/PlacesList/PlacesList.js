import React from 'react';

import PlacesListItem from '../PlacesListItem/PlacesListItem';

import './PlacesList.css';

const PlacesList = ({places, displayInfoFromListItem}) => {

  return (
    <div className="list-group places-content">
      {places.length ? places.map((place, i) =>
        <PlacesListItem 
          key={i} 
          index={i} 
          place={place} 
          displayInfoFromListItem={displayInfoFromListItem}
        />
      ) : null}
    </div>
  )
}

export default PlacesList;