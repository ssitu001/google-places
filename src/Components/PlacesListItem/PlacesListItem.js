import React from 'react';

import './PlacesListItem.css';

const PlacesListItem = ({place, index, displayInfoFromListItem}) => {

  return (
    <div 
      className={`list-group-item ${index % 2 === 0 ? 'list-group-item-alternate' : ''}`}
      onClick={() => displayInfoFromListItem(index)}>
      <h5 className="list-group-item-heading">
        {place.name}
      </h5>
      <p>{place.formatted_address}</p>
    </div>
  )
}

export default PlacesListItem;