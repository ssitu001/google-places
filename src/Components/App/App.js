import React, { Component } from 'react';

//Components
import SearchPlacesInput from '../SearchPlacesInput/SearchPlacesInput';
import PlacesList from '../PlacesList/PlacesList';

//Assets
import loadingLogo from '../../images/loading_icon.png';
import googleLogo from '../../images/google_places_logo.png';

//CSS
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = { 
      place: '',
      places: [],
      currentLocation: {lat: null, lng: null},
      markers: [],
      filters: {
        openNow: false,
        pricing: [],
      },
      filteredPlaces: [],
    };
  }

  componentDidMount() {
    this.setUserLocation();
  }

  setUserLocation() {
    if (navigator.geolocation) {
      console.log('Geolocation is supported!');
      const currentPosPromise = new Promise((resolve, reject) => {
        navigator.geolocation.getCurrentPosition(resolve, reject);
      });

      return currentPosPromise 
        .then((pos) => {
          this.setState({
            currentLocation: {
              lat: pos.coords.latitude,
              lng: pos.coords.longitude,
            }
          });
        })
        .then(() => {
          this.loadMap();
        })
        .catch((err) => {
          console.log('err::::', err);
        });
    }
    else {
      console.log('Geolocation is not supported for this Browser/OS.');
    }
  }

  loadMap() {
    const center = new window.google.maps.LatLng(this.state.currentLocation.lat, this.state.currentLocation.lng);
    
    const mapOpts = {
      zoom: 14,
      center,
    };

    this.map = new window.google.maps.Map(this.mapElementRef, mapOpts);
  }
  
  handleChange = (e) => {
    this.setState({ place: e.target.value });
  }
  
  handleSubmit = (e) => {
    e.preventDefault();
    this.searchPlaces();
  }

  searchPlaces(distance) {
    this.placesService = new window.google.maps.places.PlacesService(this.map);
    const location = new window.google.maps.LatLng(this.state.currentLocation.lat, this.state.currentLocation.lng);
    const placeToQuery = this.state.place;
    const cachedResults = JSON.parse(localStorage.getItem(placeToQuery));
    const request = {
      location: location,
      query: placeToQuery,
      radius: '100',
    }

    // validate input
    if (!placeToQuery) {
      return;
    }

    if (cachedResults) {
      this.displayResults(cachedResults)
    } else {
      this.placesService.textSearch(request, (results, status) => {
        if (status === 'OK') {
          console.log('results::', results);
          this.cacheResults(request.query, results);
        } else {
          console.log('error yo')
        }
      });
    }
  }

  cacheResults(query, results) {
    localStorage.setItem(query, JSON.stringify(results));
    this.displayResults(results);
  }

  displayResults(places) {
    if (this.state.markers.length) {
      this.removePreviousMarkers();
    }

    this.getMarkers(places);
    this.setState({ 
      places,
    });
  }

  removePreviousMarkers() {
    this.state.markers.forEach((marker) => {
      marker.setMap(null);
    });

    this.setState({
      markers: [],
    });
  }

  getMarkers(places) {
    const markers = places.map((place) => {
      return this.setMarkers(place)
    });

    this.infoWindow = new window.google.maps.InfoWindow();
    
    this.setState({ markers });
  }

  setMarkers(place, remove) {
    const contentString = `
        <div class="row">
          <div class="col-xs-12">
            <h5>${place.name}</h5>
            <p>${place.formatted_address}</p>
          </div>
        </div>
    `;

    const marker = new window.google.maps.Marker({
      position: place.geometry.location,
      title: place.name,
      map: remove ? null : this.map,
    });

    new window.google.maps.event.addListener(marker, 'click', () => {
      this.infoWindow.close();
      this.infoWindow.setContent(contentString);
      this.infoWindow.open(this.map, marker);
    })

    return marker;
  }


  displayInfoFromListItem = (idx) => {
    const { markers } = this.state;
    new window.google.maps.event.trigger(markers[idx], 'click');
  }

  // Post MVP filters
  displayAll = () => {
    this.setState({
      filters: {
        openNow: false,
        pricing: [],
      }
    });
  }

  displayOpenNow = () => {
    const { openNow, pricing } = this.state.filters;

    if (openNow) {
      this.filteredPlaces = this.state.places.filter((place) => {
        return place.opening_hours.open_now;
      });
    }

    this.setState({
      filters: {
        openNow: !openNow,
        pricing: pricing,
      }
    });

    this.filterPlaces();
  }

  filterPrice = (price) => {
    const { pricing } = this.state.filters;

    if ( pricing.includes(price) ) {
      this.setState({
        filters: {
          openNow: this.state.filters.openNow,
          pricing: pricing.filter((priceInState) => priceInState !== price),
        }
      })
    } else {
      this.setState({
        filters: {
          openNow: this.state.filters.openNow,
          pricing: [...pricing, price],
        }
      });
    }
  }

  filterPlaces() {
    const { openNow, pricing } = this.state.filters;
    let placesToFilter = this.state.places;

    if (openNow && pricing.length) {
      placesToFilter = this.state.places.filter((place) => {
        if (place) {
          if ( pricing.includes(place.price_level) && place.opening_hours.open_now) {
            return place;
          }
        }
      });
    }

    else if (openNow) {
      placesToFilter = this.state.places.filter((place) => {
        return place.opening_hours.open_now
      });
    }

    else if (pricing.length) {
      placesToFilter = this.state.places.filter((place) => {
        if ( pricing.includes(place.price_level) ) {
          return place;
        }
      });
    }
    this.setState({
      filteredPlaces: placesToFilter
    });
  }
  
  render() {
    return (
      <div className="app">
        <header className="app-header">
          <img src={googleLogo} className="app-logo-static" alt="logo" />
          <h1 className="app-title">Welcome to Places</h1>
          <SearchPlacesInput 
            handleChange={this.handleChange} 
            handleSubmit={this.handleSubmit}
          />
        </header>
        <div className="row">
          <div className={`no-padding ${this.state.places.length ? 'col-md-8' : 'col-md-12'}`}>
            <div className="map" ref={(domEl) => {this.mapElementRef = domEl}}>
              <div className="init-loading">
                <img src={loadingLogo} className="app-logo" alt="logo" />
                <h5>Initializing your location...</h5>
              </div>
            </div>
          </div>
            <div className={`no-padding ${this.state.places.length ? 'col-md-4' : ''}`}>
              <PlacesList 
                places={this.state.places}
                displayInfoFromListItem={this.displayInfoFromListItem}
              />
            </div>
          </div>
       </div>
    );
  }
}

export default App;
