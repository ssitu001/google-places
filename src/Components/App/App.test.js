import React from 'react';
import toJson from 'enzyme-to-json';

import App from './App';
import SearchPlacesInput from '../SearchPlacesInput/SearchPlacesInput';
import PlacesList from '../PlacesList/PlacesList';
import PlacesListItem from '../PlacesListItem/PlacesListItem';

describe('<App />', () => {
  test('renders App component as expected', () => {
    const component = shallow(<App />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  
  test('it contains the <SearchPlacesInput /> component', () => {
    const component = mount(<App />);
    expect(component.find(SearchPlacesInput).length).toEqual(1);
  });
  
  test('it contains the <PlacesList /> component', () => {
    const component = mount(<App />);
    expect(component.find(PlacesList).length).toEqual(1);
  });
  
  test('it should display full map when there are no places', () => {
    const component = shallow(<App />);
    component.setState({
      places: [],
    });
    
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  
  test('it should display partial map when there are places', () => {
    const component = shallow(<App />);
    component.setState({
      places: [{}, {}, {}],
    });
    
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('it should display error message when unable to find user location', () => {
    const component = shallow(<App />);
    component.setState({
      errorLocation: true,
    });

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('it should display error message when unable to find results', () => {
    const component = shallow(<App />);
    component.setState({
      errorResults: true,
    });
    
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('it should not display spinner and initializing message when unable to find results or user location', () => {
    const component = shallow(<App />);
    component.setState({
      errorResults: true,
      errorLocation: false,
    });
    
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('it should not display spinner and initializing message when unable to find results or user location', () => {
    const component = shallow(<App />);
    component.setState({
      errorResults: false,
      errorLocation: true,
    });
    
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  describe('handleChange', () => {
    test('it should set correct place on user input', () => {
      const eventMock = {
        target: {
          value: 'test',
        }
      };
      const component = shallow(<App />);
      component.instance().handleChange(eventMock);
      expect(component.state().place).toEqual('test');
    }); 
  })

  describe('handleSubmit', () => {
    test('it should call correctly handleSubmit and searchPlaces', () => {
      const component = shallow(<App />);
      const eventMock = {
        preventDefault: jest.fn(),
      };
      const searchPlacesSpy = jest.spyOn(component.instance(), 'searchPlaces');
    
      localStorage.getItem.mockReturnValueOnce(JSON.stringify([{place: 'place1'}]));
      
      component.instance().handleSubmit(eventMock);
      
      expect(eventMock.preventDefault).toHaveBeenCalled();
      expect(searchPlacesSpy).toHaveBeenCalled();
    
      searchPlacesSpy.mockReset();
      searchPlacesSpy.mockRestore();
    });
  });
});
