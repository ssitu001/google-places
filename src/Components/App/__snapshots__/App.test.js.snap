// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> it should display error message when unable to find results 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Sorry, unable to location results for your search
          </h5>
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`<App /> it should display error message when unable to find user location 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Sorry, unable to find your location
          </h5>
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`<App /> it should display full map when there are no places 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`<App /> it should display partial map when there are places 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-8"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding col-md-4"
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={
          Array [
            Object {},
            Object {},
            Object {},
          ]
        }
      />
    </div>
  </div>
</div>
`;

exports[`<App /> it should not display spinner and initializing message when unable to find results or user location 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Sorry, unable to location results for your search
          </h5>
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`<App /> it should not display spinner and initializing message when unable to find results or user location 2`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Sorry, unable to find your location
          </h5>
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`<App /> renders App component as expected 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo-static"
      src="google_places_logo.png"
    />
    <h1
      className="app-title"
    >
      Welcome to Places
    </h1>
    <SearchPlacesInput
      handleChange={[Function]}
      handleSubmit={[Function]}
    />
  </header>
  <div
    className="row"
  >
    <div
      className="no-padding col-md-12"
    >
      <div
        className="map"
      >
        <div
          className="init-loading"
        >
          <img
            alt="logo"
            className="app-logo"
            src="loading_icon.png"
          />
          <h5>
            Initializing your location...
          </h5>
        </div>
      </div>
    </div>
    <div
      className="no-padding "
    >
      <PlacesList
        displayInfoFromListItem={[Function]}
        places={Array []}
      />
    </div>
  </div>
</div>
`;
